compiler-rt/lib/hwasan/hwasan_allocator.cpp

```diff
@@ -237,7 +237,10 @@ static void *HwasanAllocate(StackTrace *stack, uptr orig_size, uptr alignment,
   if (InTaggableRegion(reinterpret_cast<uptr>(user_ptr)) &&
       atomic_load_relaxed(&hwasan_allocator_tagging_enabled) &&
       flags()->tag_in_malloc && malloc_bisect(stack, orig_size)) {
-    tag_t tag = t ? t->GenerateRandomTag() : kFallbackAllocTag;
+    tag_t tag = t ? t->GenerateRandomNonCollidingTag((uptr)user_ptr - 1,
```

Comment:
I'd rather do the same as in the Deallocate code - call GenerateRandomTag() in a loop until it returns a suitable tag. You could try lifting the (!tag) check from GenerateRandomTag to its callers to remove one nested loop, but make sure not to miss any :)

Reply:
Done, I also lifted the !tag check from GenerateRandomTag as you suggested and added the check to the other locations where the function was called and a non-zero tag is needed.

Reply:
It seems lifting the do while loop is problematic as with tagging_disabled_ set, only zero tags are generated, and the program is stuck.